File: src/Z.Collections/System.Collections.Generic.ICollection[T]/ICollection[T].RemoveWhere.cs
// Copyright (c) 2015 ZZZ Projects. All rights reserved
// Licensed under MIT License (MIT) (https://github.com/zzzprojects/Z.ExtensionMethods)
// Website: http://www.zzzprojects.com/
// Feedback / Feature Requests / Issues : http://zzzprojects.uservoice.com/forums/283927
// All ZZZ Projects products: Entity Framework Extensions / Bulk Operations / Extension Methods /Icon Library

using System;
using System.Collections.Generic;
using System.Linq;

public static partial class Extensions
{
    /// <summary>
    ///     An ICollection&lt;T&gt; extension method that removes value that satisfy the predicate.
    /// </summary>
    /// <typeparam name="T">Generic type parameter.</typeparam>
    /// <param name="this">The @this to act on.</param>
    /// <param name="predicate">The predicate.</param>
    public static void RemoveWhere<T>(this ICollection<T> @this, Func<T, bool> predicate)
    {
        List<T> list = @this.Where(predicate).ToList();
        foreach (T item in list)
        {
            @this.Remove(item);
        }
    }
}

File: src (VB.NET)/Z.ExtensionMethods.VB/Z.Collections/System.Collections.Generic.ICollection[T]/ICollection[T].RemoveWhere.vb

' Copyright (c) 2015 ZZZ Projects. All rights reserved
' Licensed under MIT License (MIT) (https://github.com/zzzprojects/Z.ExtensionMethods)
' Website: http://www.zzzprojects.com/
' Feedback / Feature Requests / Issues : http://zzzprojects.uservoice.com/forums/283927
' All ZZZ Projects products: Entity Framework Extensions / Bulk Operations / Extension Methods /Icon Library

Imports System.Collections.Generic
Imports System.Linq

Public Module Extensions_18

	''' <summary>
	'''     An ICollection&lt;T&gt; extension method that removes value that satisfy the predicate.
	''' </summary>
	''' <typeparam name="T">Generic type parameter.</typeparam>
	''' <param name="this">The @this to act on.</param>
	''' <param name="predicate">The predicate.</param>
	<System.Runtime.CompilerServices.Extension> _
	Public Sub RemoveWhere(Of T)(this As ICollection(Of T), predicate As Func(Of T, Boolean))
		Dim list As List(Of T) = this.Where(predicate).ToList()
		For Each item As T In list
			this.Remove(item)
		Next
	End Sub
End Module


